lib/sums_up/core/sum_type.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module SumsUp
+  module Core
+    # SumTypes are just modules with a meta-programmed methods to construct
+    # variants. Use SumType.build to define a new one.
+    #
+    # Each variant class must have the following constants defined:
+    #
+    # * VARIANT - Symbol name of the variant
+    # * MEMBERS - Array of Symbols which enumerate the variant's arguments
+    #
+    class SumType < Module
+      private_class_method :new
+
+      def self.build(variant_classes, &block)
+        new(&block).tap do |sum_type|
+          sum_type.const_set(:VARIANTS, variant_classes.freeze)
```

Comment:
I think I did it in this order because, if I wanted to do the meta-programmed portion first, then I believe I'd have to use an `#instance_eval` to get the block to do the right thing. I'll play around with it.